pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,321 @@
+ar.register_function("torch", "gather", lambda x, indices: x[indices])
+
+ar.register_function(
+    "torch",
+    "sqrt",
+    lambda x: _i("torch").sqrt(
+        x.to(_i("torch").float64) if x.dtype in (_i("torch").int64, _i("torch").int32) else x
+    ),
+)
+
+
+def _take_torch(tensor, indices, axis=None):
+    torch = _i("torch")
+
+    if not isinstance(indices, torch.Tensor):
+        indices = torch.as_tensor(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    if indices.ndim == 1:
+        if (indices < 0).any():
+            # index_select doesn't allow negative indices
+            dim_length = tensor.size()[0] if axis is None else tensor.shape[axis]
+            indices = torch.where(indices >= 0, indices, indices + dim_length)
+
+        return torch.index_select(tensor, dim=axis, index=indices)
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[fancy_indices]
+
+
+ar.register_function("torch", "take", _take_torch)
+
+
+def _coerce_types_torch(tensors):
+    torch = _i("torch")
+    tensors = [torch.as_tensor(t) for t in tensors]
+    dtypes = {i.dtype for i in tensors}
+
+    if len(dtypes) == 1:
+        return tensors
+
+    complex_priority = [torch.complex64, torch.complex128]
+    float_priority = [torch.float16, torch.float32, torch.float64]
+    int_priority = [torch.int8, torch.int16, torch.int32, torch.int64]
+
+    complex_type = [i for i in complex_priority if i in dtypes]
+    float_type = [i for i in float_priority if i in dtypes]
+    int_type = [i for i in int_priority if i in dtypes]
+
+    cast_type = complex_type or float_type or int_type
+    cast_type = list(cast_type)[-1]
+
+    return [t.to(cast_type) for t in tensors]
+
+
+ar.register_function("torch", "coerce", _coerce_types_torch)
+
+
+def _block_diag_torch(tensors):
+    torch = _i("torch")
+    sizes = np.array([t.shape for t in tensors])
+    res = torch.zeros(np.sum(sizes, axis=0).tolist(), dtype=tensors[0].dtype)
+
+    p = np.cumsum(sizes, axis=0)
+    ridx, cidx = np.stack([p - sizes, p]).T
+
+    for t, r, c in zip(tensors, ridx, cidx):
+        row = np.arange(*r).reshape(-1, 1)
+        col = np.arange(*c).reshape(1, -1)
+        res[row, col] = t
+
+    return res
+
+
+ar.register_function("torch", "block_diag", _block_diag_torch)
+
+
+def _scatter_element_add_torch(tensor, index, value):
+    if tensor.is_leaf:
+        tensor = tensor.clone()
+    tensor[tuple(index)] += value
+    return tensor
+
+
+ar.register_function("torch", "scatter_element_add", _scatter_element_add_torch)
+
+
+# -------------------------------- JAX --------------------------------- #
+
+
+ar.register_function("jax", "flatten", lambda x: x.flatten())
+ar.register_function(
+    "jax",
+    "take",
+    lambda x, indices, axis=None: _i("jax").numpy.take(x, indices, axis=axis, mode="wrap"),
+)
+ar.register_function("jax", "coerce", lambda x: x)
+ar.register_function("jax", "to_numpy", lambda x: x)

```

Comment:
This you can have as `lambda x: onp.array(x)` If x is a traced array, then this will throw an error (like it should).

Reply:
oh, good idea
```suggestion
ar.register_function("jax", "to_numpy", lambda x: onp.array(x))
```

Reply:
Unresolved this, seems like the suggestion was eventually not considered, any reason for that?

Reply:
@antalszava I had to revert it 🙁 Unfortunately there are unrelated parts of the codebase, including `default.qubit.jax`, that apply `to_numpy()`, and this breaks. This is partly because other frameworks (TF, Torch, Autograd) allow conversion to numpy arrays even though it breaks autodiff, and this assumption is used a lot.